### DISCUSSION TITLE
Multiple tenants with multiple instances

### DISCUSSION THREAD

**Discussion by Gerald-Gmainer** [supabase/supabase#21433] [Questions]
Our client has a supabase instance with about 130 tenants. Each tenant has about 50 users. Each users inserts, updates, etc. about 15 times per day. Overall 120k rest requests per day. There is a webfront and mobile app.

Client wants to seperate the instances due to performance. So far, we have no table partitions or so. 99% data is stored in 2 tables. Partitions would solve the performance problems at the moment, but the client wants something more futureproof.

So we would like to separate the supabase instances, but keep 1 single endpoints for web and mobile app. I guess we need 1 main instance for login and other stuff. And some supabase instances for tenants. Maybe 20 tenants per instance.

I have no clue, how to setup this approach. My main concerns
* how to do the login (how to login to users tenant instance; how does the main instance know the users of all instances; how does the main instance interact with all instances)
* how to publish updates all instances (DB migration, edge function)
* what about the existing mobile app; is there a way that old app versions work with the new login

Any advice and suggestions are welcome

**Comment by GaryAustin1** [comment]
Sounds like the new read replicas will be the future proof way to go for something like this.
https://supabase.com/blog/introducing-read-replicas

Maintaining multiple instances with the same schemas, things staying in sync, dealing with auth across all instances seems painful.  You can set the jwt secret the same for all instances at least and share the user jwt from a single auth point.

**Reply by Gerald-Gmainer** [reply]
Thank you for your help. I am wondering about refreshing tokens, after altering the jwt. Will it cause problems?

**Reply by GaryAustin1** [reply]
You can't use the access token/refresh token on multiple devices.   The refresh token can only be used once.  But you can use that token to access multiple instances if they have the same secret.

**Comment by kevcodez** [comment]
Doesn't really sound like you need to separate instances (which would increase complexity by manifold), the volume you mentioned isn't bad and a single instance should be able to handle that without issues.

Have you looked into query performance to identify slow running queries? Have you looked into bumping up your compute size?

**Reply by Gerald-Gmainer** [reply]
Thank you. The querys are slow, but can be improved. My client still wants to split the instances for more future-proofness
